enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.ts

```diff
@@ -329,7 +425,170 @@ const raiseError = (message: string) => {
   throw new MetabaseError("EMBED_ERROR", message);
 };
 
-const warn = (...messages: unknown[]) =>
-  console.warn("[metabase.embed.warning]", ...messages);
+function createCustomElement<Arr extends readonly string[]>(
+  tagName: string,
+  attributeNames: Arr,
+) {
+  class CustomEmbedElement extends HTMLElement {
+    private _embed: MetabaseEmbed | null = null;
+
+    static get observedAttributes() {
+      return attributeNames as readonly string[];
+    }
+
+    addEventListener(
+      type: string,
+      listener: EventListenerOrEventListenerObject | SdkIframeEmbedEventHandler,
+      options?: boolean | AddEventListenerOptions,
+    ): void {
+      if (type === "ready") {
+        // Forward Metabase SDK specific events to the underlying embed instance.
+        this._embed?.addEventListener(
+          type as SdkIframeEmbedEvent["type"],
+          listener as SdkIframeEmbedEventHandler,
+        );
+        return;
+      }
+
+      // Fall back to the native HTMLElement event mechanism for all other events.
+      super.addEventListener(
+        type,
+        listener as EventListenerOrEventListenerObject,
+        options,
+      );
+    }
+
+    removeEventListener(
+      type: string,
+      listener: EventListenerOrEventListenerObject | SdkIframeEmbedEventHandler,
+      options?: boolean | EventListenerOptions,
+    ): void {
+      if (type === "ready") {
+        this._embed?.removeEventListener(
+          type as SdkIframeEmbedEvent["type"],
+          listener as SdkIframeEmbedEventHandler,
+        );
+        return;
+      }
+
+      super.removeEventListener(
+        type,
+        listener as EventListenerOrEventListenerObject,
+        options,
+      );
+    }
+
+    updateSettings(settings: Partial<SdkIframeEmbedSettings>) {
+      this._embed?.updateSettings(settings);
+    }
+
+    connectedCallback() {
+      this.style.display = "block";
+
+      if (this._embed) {
+        // already initialised
+        return;
+      }
+
+      // we need to copy the settings, we don't want to accidentally mutate the shared object
+      // TODO: do deep copy to be sure we're not sharing the theme
```

Comment:
We can use [structuredClone](https://developer.mozilla.org/en-US/docs/Web/API/Window/structuredClone) here